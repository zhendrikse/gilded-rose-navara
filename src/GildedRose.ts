'use strict';

import internal from "stream";

export class Item {
  name: string = "";
  sellIn: number = 0;
  quality: number = 0;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }

  public incrementQuality() {
    if (this.quality < 50)
      this.quality = this.quality + 1;
  }

  public decrementQuality() {
    if (this.quality > 0)
      this.quality = this.quality - 1;
  }

  public update() {
    this.decrementQuality();
    this.sellIn = this.sellIn - 1;
    this.decrementQuality();
  }

  public toString(): string {
    return "name: " + this.name + ", sellIn: " + this.sellIn + ", quality: " + this.quality
  }
}

export class AgedBrie extends Item {
  constructor(sellIn: number, quality: number) {
    super("Aged Brie", sellIn, quality);
  }

  public update() {
    this.incrementQuality();
    this.sellIn = this.sellIn - 1;
    this.incrementQuality();
  }
}

export class Sulfuras extends Item {
  constructor(sellIn: number, quality: number) {
    super("Sulfuras, Hand of Ragnaros", sellIn, quality);
  }

  public update() {
  }
}

export class BackstagePass extends Item {
  constructor(sellIn: number, quality: number) {
    super("Backstage passes to a TAFKAL80ETC concert", sellIn, quality);
  }

  public update() {
    if (this.quality < 50)
      this.quality = this.quality + 1;
    if (this.sellIn < 11)
      if (this.quality < 50)
        this.quality = this.quality + 1;
    if (this.sellIn < 6)
      if (this.quality < 50)
        this.quality = this.quality + 1;
    
    this.sellIn = this.sellIn - 1;
    
    if (this.sellIn < 0)
      this.quality = 0;
  }
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  public updateQuality(): Array<Item> {
    this.items.forEach(item => item.update());
    return this.items;
  }
}