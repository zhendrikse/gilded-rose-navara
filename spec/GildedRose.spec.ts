import { Item, GildedRose } from '../src/GildedRose';
import { verify } from "approvals/lib/Providers/Jest/JestApprovals";
import { configure } from "approvals/lib/config";
import { JestReporter } from "approvals/lib/Providers/Jest/JestReporter";

function convert_items_to_string(items = [] as Array<Item>) {
  let items_as_string = items.map((item) => item.toString() + "\n")
  return items_as_string.reduce(
    (accumulator, currentValue) => accumulator + currentValue,
    "",
  );
}

describe('Gilded Rose', () => {
  beforeAll(() => {
    configure({
      reporters: [new JestReporter()],
    });
  });

  it('updates a foo item', () => {
    const items = [
      new Item("Foo", 0, 0)
    ];
    const gildedRose = new GildedRose(items);
    const updated_items = gildedRose.updateQuality();
    verify(convert_items_to_string(updated_items));
  });
});
