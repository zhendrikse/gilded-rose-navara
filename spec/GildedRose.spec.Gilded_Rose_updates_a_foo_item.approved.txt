name: Foo, sellIn: -1, quality: 0
